pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -681,5 +862,66 @@ void topicTerminated() {
         }
     }
 
+    /**
+     * Commit a transaction.
+     *
+     * @param txnID         {@link TxnID} to identify the transaction.
+     * @param properties    Additional user-defined properties that can be associated with a particular cursor position.
+     * @throws IllegalArgumentException
+     * @throws TransactionConflictException
+     */
+    public synchronized void commitTxn(TxnID txnID, Map<String,Long> properties) throws TransactionConflictException {
+        if (!this.pendingAckMessagesMap.containsKey(txnID)) {
+            String errorMsg = "[" + topicName + "][" + subName + "] Transaction with id:" + txnID.getMostSigBits()
+                            + txnID.getLeastSigBits() + " not found.";
+            log.error(errorMsg);
+            throw new IllegalArgumentException(errorMsg);
+        }
+
+        // This shouldn't happen.
+        if (null != this.pendingCumulativeAckMessage && txnID != this.txnID) {
+            String errorMsg = "[" + topicName + "][" + subName + "] Committing transaction:" + txnID.getMostSigBits()
+                    + txnID.getLeastSigBits() + "but another transaction:" + this.txnID.getMostSigBits()
+                    + this.txnID.getLeastSigBits() + " cumulative acknowledged.";
+            log.error(errorMsg);
+            throw new TransactionConflictException(errorMsg);
+        }
+
+        ConcurrentOpenHashSet<Position> pendingAckMessage = this.pendingAckMessagesMap.remove(txnID);
+        List<Position> positions = pendingAckMessage.values();
+        // Materialize all single acks.
+        cursor.asyncDelete(positions, deleteCallback, positions);
+        positions.forEach(position -> this.pendingAckMessages.remove(position));
+
+        // Materialize cumulative ack.
+        cursor.asyncMarkDelete(this.pendingCumulativeAckMessage, (null == properties)?
+                Collections.emptyMap() : properties, markDeleteCallback, this.pendingCumulativeAckMessage);
+
+        // Reset txdID and position for cumulative ack.
+        TXNID_UPDATER.set(this, null);
+        POSITION_UPDATER.set(this, null);
+    }
+
+    /**
+     * Abort a transaction.
+     *
+     * @param txnID  {@link TxnID} to identify the transaction.
+     * @throws IllegalArgumentException
+     */
+    public synchronized void abortTxn(TxnID txnID) {
```

Comment:
similar comments as commitTxn - return `CompletableFuture<Void>`.

In this PR you can handle in-memory pending_ack state, but in subsequent PRs, we need to look into persistenting those states into cursor ledger for handling broker crashes.